File: composey/models/gcp.py
"""GCP resource models for Terraform generation.

These models represent GCP resources that the compiler can generate.
Each model maps to a Google Terraform provider resource.
"""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class CloudRunService(BaseModel):
    """Cloud Run service resource."""

    model_config = ConfigDict(extra="forbid")

    name: str
    location: str
    project_id: str

    # Container configuration
    template: Dict[str, Any] = Field(
        default_factory=dict,
        description="Service template with containers, scaling, etc.",
    )

    # Traffic configuration
    traffic: Optional[List[Dict[str, Any]]] = Field(
        default=None,
        description="Traffic routing configuration",
    )

    # VPC access
    vpc_access: Optional[Dict[str, Any]] = Field(
        default=None,
        description="VPC connector configuration",
    )

    # IAM
    autogenerate_revision_name: bool = True

    # Ingress settings
    ingress: str = Field(
        default="all",
        description="Ingress settings: all, internal, internal-and-cloud-load-balancing",
    )

    # URL handling
    depends_on: Optional[List[str]] = None


class CloudRunServiceIamMember(BaseModel):
    """IAM binding for Cloud Run service."""

    model_config = ConfigDict(extra="forbid")

    service: str  # Reference to cloud run service
    location: str
    project_id: str
    role: str
    member: str


class CloudSqlInstance(BaseModel):
    """Cloud SQL instance resource."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str
    region: str

    # Database version
    database_version: str = Field(
        default="POSTGRES_14",
        description="POSTGRES_14, MYSQL_8_0, etc.",
    )

    # Tier / machine type
    tier: str = Field(
        default="db-f1-micro",
        description="Machine type (db-f1-micro, db-g1-small, etc.)",
    )

    # Storage
    storage_auto_resize: bool = True
    storage_auto_resize_limit: int = Field(default=100, description="GB")

    # High availability
    availability_type: str = Field(
        default="ZONAL",
        description="ZONAL or REGIONAL",
    )

    # Backups
    backup_configuration: Optional[Dict[str, Any]] = Field(
        default_factory=lambda: {"enabled": True, "start_time": "03:00"},
    )

    # IP configuration
    ip_configuration: Dict[str, Any] = Field(
        default_factory=lambda: {
            "ipv4_enabled": True,
            "private_network": None,
        },
    )

    # Root password
    root_password: Optional[str] = None

    # Database to create
    database_name: Optional[str] = None


class CloudSqlDatabase(BaseModel):
    """Database within a Cloud SQL instance."""

    model_config = ConfigDict(extra="forbid")

    name: str
    instance: str  # Reference to cloud_sql_instance
    project_id: str


class RedisInstance(BaseModel):
    """Memorystore Redis instance resource."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str
    region: str

    # Tier
    tier: str = Field(
        default="BASIC",
        description="BASIC, STANDARD_HA",
    )

    # Memory size
    memory_size_gb: int = Field(default=1, ge=1, le=300)

    # Redis version
    redis_version: str = Field(default="REDIS_6_X")

    # Network
    authorized_network: Optional[str] = None
    connect_mode: str = "DIRECT_PEERING"


class StorageBucket(BaseModel):
    """Cloud Storage bucket resource."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str
    location: str = Field(default="US")

    # Storage class
    storage_class: str = Field(default="STANDARD")

    # Versioning
    versioning: Dict[str, bool] = Field(default_factory=lambda: {"enabled": False})

    # Lifecycle rules
    lifecycle_rule: Optional[List[Dict[str, Any]]] = None

    # Uniform bucket-level access
    uniform_bucket_level_access: bool = True

    # Force destroy for cleanup
    force_destroy: bool = False


class StorageBucketIamMember(BaseModel):
    """IAM binding for Storage bucket."""

    model_config = ConfigDict(extra="forbid")

    bucket: str
    role: str
    member: str


class VpcConnector(BaseModel):
    """VPC Access connector for Cloud Run."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str
    region: str

    # Network config
    network: str = "default"
    ip_cidr_range: str = Field(default="10.8.0.0/28")

    # Machine type
    machine_type: str = Field(default="f1-micro")

    # Min/max instances
    min_instances: int = 2
    max_instances: int = 10


class GlobalAddress(BaseModel):
    """Global static IP address for load balancing."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str

    address_type: str = "EXTERNAL"
    ip_version: str = "IPV4"


class ComputeManagedSslCertificate(BaseModel):
    """Managed SSL certificate for load balancing."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str

    managed: Dict[str, List[str]] = Field(default_factory=lambda: {"domains": []})


class ComputeUrlMap(BaseModel):
    """URL map for load balancing."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str

    default_service: str  # Reference to backend service
    host_rule: Optional[List[Dict[str, Any]]] = None
    path_matcher: Optional[List[Dict[str, Any]]] = None


class ComputeBackendService(BaseModel):
    """Backend service for load balancing."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str

    backend: List[Dict[str, Any]]

    # Cloud Run can be a backend
    # Using NEG (Network Endpoint Group)


class ComputeTargetHttpsProxy(BaseModel):
    """HTTPS proxy for load balancing."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str

    url_map: str
    ssl_certificates: List[str]


class ComputeForwardingRule(BaseModel):
    """Forwarding rule for load balancing."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str

    target: str  # Reference to HTTPS proxy
    ip_address: Optional[str] = None
    port_range: str = "443"


class ComputeRegionNetworkEndpointGroup(BaseModel):
    """Network endpoint group for Cloud Run."""

    model_config = ConfigDict(extra="forbid")

    name: str
    project_id: str
    region: str

    network_endpoint_type: str = "SERVERLESS"
    cloud_run: Dict[str, str]  # service name


class SecretManagerSecret(BaseModel):
    """Secret Manager secret resource."""

    model_config = ConfigDict(extra="forbid")

    secret_id: str
    project_id: str

    replication: Dict[str, Any] = Field(default_factory=lambda: {"automatic": {}})


class SecretManagerSecretVersion(BaseModel):
    """Secret version resource."""

    model_config = ConfigDict(extra="forbid")

    secret: str  # Reference to secret
    secret_data: str


class GcpResources(BaseModel):
    """A registry of the GCP resources our compiler supports."""

    # Cloud Run
    google_cloud_run_service: Dict[str, CloudRunService] = Field(default_factory=dict)
    google_cloud_run_service_iam_member: Dict[str, CloudRunServiceIamMember] = Field(
        default_factory=dict
    )

    # Cloud SQL
    google_sql_database_instance: Dict[str, CloudSqlInstance] = Field(
        default_factory=dict
    )
    google_sql_database: Dict[str, CloudSqlDatabase] = Field(default_factory=dict)

    # Memorystore (Redis)
    google_redis_instance: Dict[str, RedisInstance] = Field(default_factory=dict)

    # Cloud Storage
    google_storage_bucket: Dict[str, StorageBucket] = Field(default_factory=dict)
    google_storage_bucket_iam_member: Dict[str, StorageBucketIamMember] = Field(
        default_factory=dict
    )

    # VPC
    google_vpc_access_connector: Dict[str, VpcConnector] = Field(default_factory=dict)

    # Load Balancing (for CDN/custom domains)
    google_compute_global_address: Dict[str, GlobalAddress] = Field(
        default_factory=dict
    )
    google_compute_managed_ssl_certificate: Dict[str, ComputeManagedSslCertificate] = (
        Field(default_factory=dict)
    )
    google_compute_region_network_endpoint_group: Dict[
        str, ComputeRegionNetworkEndpointGroup
    ] = Field(default_factory=dict)
    google_compute_backend_service: Dict[str, ComputeBackendService] = Field(
        default_factory=dict
    )
    google_compute_url_map: Dict[str, ComputeUrlMap] = Field(default_factory=dict)
    google_compute_target_https_proxy: Dict[str, ComputeTargetHttpsProxy] = Field(
        default_factory=dict
    )
    google_compute_forwarding_rule: Dict[str, ComputeForwardingRule] = Field(
        default_factory=dict
    )

    # Secret Manager
    google_secret_manager_secret: Dict[str, SecretManagerSecret] = Field(
        default_factory=dict
    )
    google_secret_manager_secret_version: Dict[str, SecretManagerSecretVersion] = Field(
        default_factory=dict
    )

    # Docker provider resources
    docker_image: Dict[str, Any] = Field(default_factory=dict)
    docker_registry_image: Dict[str, Any] = Field(default_factory=dict)

    # Random resources for passwords
    random_password: Dict[str, Any] = Field(default_factory=dict)

File: cloudcompose-go/internal/models/gcp.go
package models

// GCP resource models.
//
// Field names and JSON tags match Terraform's own attribute names exactly,
// since these marshal straight into Terraform's JSON syntax.
//
// Unlike AWS/Azure, this has essentially no test surface to verify
// against, so it's verified with lighter sanity-checking only.

type CloudRunService struct {
	Name      string `json:"name"`
	Location  string `json:"location"`
	ProjectID string `json:"project_id"`

	Template  *CloudRunTemplate `json:"template,omitempty"`
	Traffic   []CloudRunTraffic `json:"traffic,omitempty"`
	VpcAccess any               `json:"vpc_access,omitempty"`

	AutogenerateRevisionName bool     `json:"autogenerate_revision_name"`
	Ingress                  string   `json:"ingress"`
	DependsOn                []string `json:"depends_on,omitempty"`
}

func NewCloudRunService() CloudRunService {
	return CloudRunService{AutogenerateRevisionName: true, Ingress: "all"}
}

// CloudRunTemplate is google_cloud_run_service's `template` block: the
// container spec plus autoscaling annotations.
type CloudRunTemplate struct {
	Spec     CloudRunSpec         `json:"spec"`
	Metadata CloudRunTemplateMeta `json:"metadata"`
}

type CloudRunSpec struct {
	Containers         []CloudRunContainer `json:"containers"`
	ServiceAccountName string              `json:"service_account_name,omitempty"`
}

type CloudRunContainer struct {
	Image     string                  `json:"image"`
	Command   []string                `json:"command,omitempty"`
	Env       []CloudRunEnvVar        `json:"env,omitempty"`
	Resources CloudRunContainerLimits `json:"resources"`
}

type CloudRunEnvVar struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type CloudRunContainerLimits struct {
	Limits CloudRunResourceLimits `json:"limits"`
}

type CloudRunResourceLimits struct {
	CPU    string `json:"cpu"`
	Memory string `json:"memory"`
}

type CloudRunTemplateMeta struct {
	Annotations map[string]string `json:"annotations,omitempty"`
}

type CloudRunTraffic struct {
	Percent        int  `json:"percent"`
	LatestRevision bool `json:"latest_revision"`
}

type CloudRunServiceIamMember struct {
	Service   string `json:"service"`
	Location  string `json:"location"`
	ProjectID string `json:"project_id"`
	Role      string `json:"role"`
	Member    string `json:"member"`
}

type CloudSqlInstance struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`
	Region    string `json:"region"`

	DatabaseVersion string `json:"database_version"`
	Tier            string `json:"tier"`

	StorageAutoResize      bool `json:"storage_auto_resize"`
	StorageAutoResizeLimit int  `json:"storage_auto_resize_limit"`

	AvailabilityType string `json:"availability_type"`

	BackupConfiguration *CloudSqlBackupConfiguration `json:"backup_configuration,omitempty"`
	IpConfiguration     CloudSqlIPConfiguration      `json:"ip_configuration"`

	RootPassword *string `json:"root_password,omitempty"`
	DatabaseName *string `json:"database_name,omitempty"`
}

type CloudSqlBackupConfiguration struct {
	Enabled   bool   `json:"enabled"`
	StartTime string `json:"start_time"`
}

type CloudSqlIPConfiguration struct {
	Ipv4Enabled    bool    `json:"ipv4_enabled"`
	PrivateNetwork *string `json:"private_network,omitempty"`
}

// NewCloudSqlInstance returns a CloudSqlInstance with its scalar
// defaults set.
func NewCloudSqlInstance() CloudSqlInstance {
	return CloudSqlInstance{
		DatabaseVersion:        "POSTGRES_14",
		Tier:                   "db-f1-micro",
		StorageAutoResize:      true,
		StorageAutoResizeLimit: 100,
		AvailabilityType:       "ZONAL",
	}
}

type CloudSqlDatabase struct {
	Name      string `json:"name"`
	Instance  string `json:"instance"`
	ProjectID string `json:"project_id"`
}

type RedisInstance struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`
	Region    string `json:"region"`

	Tier         string `json:"tier"`
	MemorySizeGb int    `json:"memory_size_gb"`
	RedisVersion string `json:"redis_version"`

	AuthorizedNetwork *string `json:"authorized_network,omitempty"`
	ConnectMode       string  `json:"connect_mode"`
}

func NewRedisInstance() RedisInstance {
	return RedisInstance{
		Tier:         "BASIC",
		MemorySizeGb: 1,
		RedisVersion: "REDIS_6_X",
		ConnectMode:  "DIRECT_PEERING",
	}
}

type StorageBucket struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`
	Location  string `json:"location"`

	StorageClass string                  `json:"storage_class"`
	Versioning   StorageBucketVersioning `json:"versioning"`

	LifecycleRule []map[string]any `json:"lifecycle_rule,omitempty"`

	UniformBucketLevelAccess bool `json:"uniform_bucket_level_access"`
	ForceDestroy             bool `json:"force_destroy"`
}

type StorageBucketVersioning struct {
	Enabled bool `json:"enabled"`
}

func NewStorageBucket() StorageBucket {
	return StorageBucket{
		Location:                 "US",
		StorageClass:             "STANDARD",
		UniformBucketLevelAccess: true,
	}
}

type StorageBucketIamMember struct {
	Bucket string `json:"bucket"`
	Role   string `json:"role"`
	Member string `json:"member"`
}

type VpcConnector struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`
	Region    string `json:"region"`

	Network     string `json:"network"`
	IpCidrRange string `json:"ip_cidr_range"`
	MachineType string `json:"machine_type"`

	MinInstances int `json:"min_instances"`
	MaxInstances int `json:"max_instances"`
}

func NewVpcConnector() VpcConnector {
	return VpcConnector{
		Network:      "default",
		IpCidrRange:  "10.8.0.0/28",
		MachineType:  "f1-micro",
		MinInstances: 2,
		MaxInstances: 10,
	}
}

type GlobalAddress struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`

	AddressType string `json:"address_type"`
	IpVersion   string `json:"ip_version"`
}

func NewGlobalAddress() GlobalAddress {
	return GlobalAddress{AddressType: "EXTERNAL", IpVersion: "IPV4"}
}

type ComputeManagedSslCertificate struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`

	Managed any `json:"managed"`
}

func NewComputeManagedSslCertificate() ComputeManagedSslCertificate {
	return ComputeManagedSslCertificate{}
}

type ComputeUrlMap struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`

	DefaultService string           `json:"default_service"`
	HostRule       []map[string]any `json:"host_rule,omitempty"`
	PathMatcher    []map[string]any `json:"path_matcher,omitempty"`
}

type ComputeBackendService struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`

	Backend []map[string]any `json:"backend"`
}

type ComputeTargetHttpsProxy struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`

	UrlMap          string   `json:"url_map"`
	SslCertificates []string `json:"ssl_certificates"`
}

type ComputeForwardingRule struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`

	Target    string  `json:"target"`
	IpAddress *string `json:"ip_address,omitempty"`
	PortRange string  `json:"port_range"`
}

func NewComputeForwardingRule() ComputeForwardingRule {
	return ComputeForwardingRule{PortRange: "443"}
}

type ComputeRegionNetworkEndpointGroup struct {
	Name      string `json:"name"`
	ProjectID string `json:"project_id"`
	Region    string `json:"region"`

	NetworkEndpointType string            `json:"network_endpoint_type"`
	CloudRun            map[string]string `json:"cloud_run"`
}

func NewComputeRegionNetworkEndpointGroup() ComputeRegionNetworkEndpointGroup {
	return ComputeRegionNetworkEndpointGroup{NetworkEndpointType: "SERVERLESS"}
}

type SecretManagerSecret struct {
	SecretID  string `json:"secret_id"`
	ProjectID string `json:"project_id"`

	Replication any `json:"replication"`
}

func NewSecretManagerSecret() SecretManagerSecret {
	return SecretManagerSecret{}
}

type SecretManagerSecretVersion struct {
	Secret     string `json:"secret"`
	SecretData string `json:"secret_data"`
}

// GcpResources is a registry of the GCP resources the compiler supports.
type GcpResources struct {
	CloudRunService          map[string]CloudRunService          `json:"google_cloud_run_service,omitempty"`
	CloudRunServiceIamMember map[string]CloudRunServiceIamMember `json:"google_cloud_run_service_iam_member,omitempty"`

	SqlDatabaseInstance map[string]CloudSqlInstance `json:"google_sql_database_instance,omitempty"`
	SqlDatabase         map[string]CloudSqlDatabase `json:"google_sql_database,omitempty"`

	RedisInstance map[string]RedisInstance `json:"google_redis_instance,omitempty"`

	StorageBucket          map[string]StorageBucket          `json:"google_storage_bucket,omitempty"`
	StorageBucketIamMember map[string]StorageBucketIamMember `json:"google_storage_bucket_iam_member,omitempty"`

	VpcAccessConnector map[string]VpcConnector `json:"google_vpc_access_connector,omitempty"`

	ComputeGlobalAddress              map[string]GlobalAddress                     `json:"google_compute_global_address,omitempty"`
	ComputeManagedSslCertificate      map[string]ComputeManagedSslCertificate      `json:"google_compute_managed_ssl_certificate,omitempty"`
	ComputeRegionNetworkEndpointGroup map[string]ComputeRegionNetworkEndpointGroup `json:"google_compute_region_network_endpoint_group,omitempty"`
	ComputeBackendService             map[string]ComputeBackendService             `json:"google_compute_backend_service,omitempty"`
	ComputeUrlMap                     map[string]ComputeUrlMap                     `json:"google_compute_url_map,omitempty"`
	ComputeTargetHttpsProxy           map[string]ComputeTargetHttpsProxy           `json:"google_compute_target_https_proxy,omitempty"`
	ComputeForwardingRule             map[string]ComputeForwardingRule             `json:"google_compute_forwarding_rule,omitempty"`

	SecretManagerSecret        map[string]SecretManagerSecret        `json:"google_secret_manager_secret,omitempty"`
	SecretManagerSecretVersion map[string]SecretManagerSecretVersion `json:"google_secret_manager_secret_version,omitempty"`

	DockerImage         map[string]DockerImage         `json:"docker_image,omitempty"`
	DockerRegistryImage map[string]DockerRegistryImage `json:"docker_registry_image,omitempty"`

	RandomPassword map[string]RandomPassword `json:"random_password,omitempty"`
}

// NewGcpResources returns a GcpResources with every map initialized.
func NewGcpResources() *GcpResources {
	return &GcpResources{
		CloudRunService:                   map[string]CloudRunService{},
		CloudRunServiceIamMember:          map[string]CloudRunServiceIamMember{},
		SqlDatabaseInstance:               map[string]CloudSqlInstance{},
		SqlDatabase:                       map[string]CloudSqlDatabase{},
		RedisInstance:                     map[string]RedisInstance{},
		StorageBucket:                     map[string]StorageBucket{},
		StorageBucketIamMember:            map[string]StorageBucketIamMember{},
		VpcAccessConnector:                map[string]VpcConnector{},
		ComputeGlobalAddress:              map[string]GlobalAddress{},
		ComputeManagedSslCertificate:      map[string]ComputeManagedSslCertificate{},
		ComputeRegionNetworkEndpointGroup: map[string]ComputeRegionNetworkEndpointGroup{},
		ComputeBackendService:             map[string]ComputeBackendService{},
		ComputeUrlMap:                     map[string]ComputeUrlMap{},
		ComputeTargetHttpsProxy:           map[string]ComputeTargetHttpsProxy{},
		ComputeForwardingRule:             map[string]ComputeForwardingRule{},
		SecretManagerSecret:               map[string]SecretManagerSecret{},
		SecretManagerSecretVersion:        map[string]SecretManagerSecretVersion{},
		DockerImage:                       map[string]DockerImage{},
		DockerRegistryImage:               map[string]DockerRegistryImage{},
		RandomPassword:                    map[string]RandomPassword{},
	}
}
